### PR TITLE
Huge cleanup, worthy of new minor version (breaking changes)

### PR DESCRIPTION
**Pull request by cozyGalvinism**
Hello! I am very thankful for this port! It makes storing large numbers much easier and hassle-free. Yet I came across some issues and improvements, which I have worked on. A full list can be found in the changelog, but among the changes were the following:

- `to_exponential()` was broken and lead to an endless loop for precisions above 2 
- `to_fixed()` panicked when the number had no significant digits (like `10.0`).
- The `Display` trait was specified along with an additional `to_string()` function (which shadowed the `to_string()` of `Display`.
- Default implementation of `Add`, `Sub`,  `Mul` and `Div` were done via borrow-to-borrow, which wasn't that bad, but in general it's safer to implement it on non-borrowed types primarily and then simply dereference in the borrow implementation.
- You couldn't format a `Decimal` struct with precision (`formatter.precision()`), like `format!("{:.4}", decimal)`
- You couldn't directly convert from `f64` (or any number type in fact) to Decimal

I have also implemented a slight workaround for reading big numbers into `Decimal`. Before, reading `1e308` into `Decimal` caused the mantissa to become a number very close to 1 (but not quite 1), now that doesn't happen anymore. It gives up a bit of mantissa precision for accurately storing big numbers. Smaller numbers don't have the issue.

Anyways, I really love this port. It's perfect for implementing into custom idle games! Personally I will modify this a little further for use within the Godot Engine (mainly conversion from and to Variant, maybe this could become a crate feature? It's not a terrible lot of code to enable).

### CODE REVIEW

**Comment by redfire75369** [comment]
Hello there! I am extremely grateful for this PR, especially since I had made a lot of mistakes for my first project in Rust. All your changes seem reasonable to me.

It seems like you included formatting changes in this PR, can you please revert those? The formatting changes happened to change the tabs to spaces. This PR can be merged after those, and after some minor changes I have been considering prior to this, v0.3.0 will be released.

It was certainly my mistake that I did not include a `rustfmt.toml` (My usual configuration includes `hard_tabs = true`). However, I had included `indent_style = tab` in them, which indicates my preference for tabs instead of spaces.

**Comment by cozyGalvinism** [comment]
Essentially I ran rust-analyzer's formatting feature over them, I converted the spaces to tabs now, though removing the entire formatting is a bit tiresome. If wanted, I can still do that though.

Other than that, there really are THAT many changes to it. I spent the last 2 days working on it for using it in a game I'm working on c: I should also mention that I included some documentation for what most of the functions do, which also increased the commit size.

**Comment by redfire75369** [comment]
I don't mind the large commits, it's just that the changes to spaces also made it difficult to look through the changes in the commit. I'll look through it again now. Thanks for the quick response!

File: src/lib.rs [hunk 1]
```diff
@@ -871,12 +923,13 @@ impl Decimal {
 	}
 
 	pub fn log2(&self) -> f64 {
-		3.32192809488736234787 * self.log10()
+		LOG2_10
```
Comment (redfire75369) [inline src/lib.rs]: `LOG2_10` is a constant. `Decimal::log2` should return the logarithm of base 2 of the current value. You probably meant to multiply it by `self.log10()`.

Comment (cozyGalvinism) [reply src/lib.rs:926]: Actually no ^^ Let me explain the change to that line:

The floating point precision of 3.32192809488736234787 is too large to fit in an `f64`, which is truncated down to 3.321928094887362 which in turn, when multiplied with `log10`, produces a very close approximation to the `LOG2_10` constant ^^ [It's explained here](https://rust-lang.github.io/rust-clippy/master/index.html#approx_constant)

`LOG2_10` is the value of `log2(10)`.

Comment (cozyGalvinism) [reply src/lib.rs]: Scratch my previous reply, I think I see what you mean.

Comment (cozyGalvinism) [reply src/lib.rs]: Yea, you're right, clippy was too overzealous here. Gonna amend that

Comment (cozyGalvinism) [reply src/lib.rs]: Fixed in dfa2f68

**Pull request merged**